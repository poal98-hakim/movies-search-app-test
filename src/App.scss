.wrapper-filter{
    padding: 10px;
}

.no-results-wrapper{
    color: red;
    font-weight: 900;
    font-size: 3em;
    width: 70%;
    height: 2em;
}