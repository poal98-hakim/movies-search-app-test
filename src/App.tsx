import { useState } from 'react';
import './App.scss';
import { IMovie } from './models/Movie';
import SearchSection from './components/SearchSection/SearchSection';
import MoviesSection from './components/MoviesSection/MoviesSection';
import MainContext from './contexts/MainContext';
import useDidMountEffect from './hooks/useDidMountEffect';
import { getAPIError } from './api/utils/methods';
import MoviesAPI from './api/movies';

const App: React.VFC = () => {

  const [movies, setMovies] = useState<IMovie[]>([]);
  const [search, setSearch] = useState<string>("");
  const [apiKey, setApiKey] = useState<string>("");
  const [page, setPage] = useState<number>(0);
  const [totalResults, setTotalResults] = useState<number>();
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [error, setError] = useState<string>("");
  const [hasMore, setHasMore] = useState<boolean>(false);
  
  const resetValues = (offset?: number) : void => {
    setIsLoading(true);
    setError("");
    setMovies([]);
    if(offset === undefined){
      setPage(0);
    }
  }

  // Update search results based on the keyword and api key inputs
  const searchMovies = async (query: string, apiKey: string, offset?: number) : Promise<void> => {
    resetValues(offset);
    const { data, errorStatus } = await MoviesAPI.getMovieReviews(apiKey, {query, offset}, () => setIsLoading(false));
    if(errorStatus){
      const _error = getAPIError(errorStatus);
      setError(_error);
      return;
    }
    const {results, num_results, has_more} = data;
    setMovies(results);
    setTotalResults(num_results);
    setHasMore(has_more);
  }

  useDidMountEffect(() => { //run on each update of page (not on mount)
    searchMovies(search, apiKey, page);
  }, [page])

  
  const result = () : JSX.Element | null => {
    if(totalResults || isLoading){
      return <MoviesSection movies={movies} hasMore={hasMore} isLoading={isLoading} />;
    }
    else if(totalResults === 0){
      return <div className="no-results-wrapper flex f-justify-center f-align-center">No Results</div>
    }
    return null;
  }

  return (
    <MainContext.Provider value={{search, apiKey, page, setSearch, setApiKey, setPage}}>
      <main>
        <div className="wrapper-filter">
          <SearchSection onSearch={(keyword, apiKey) => searchMovies(keyword, apiKey)} totalResults={totalResults} error={error} />
        </div>
        <div className="flex f-justify-center">
          {result()}
        </div>
      </main>
    </MainContext.Provider>
  );
}

export default App;
