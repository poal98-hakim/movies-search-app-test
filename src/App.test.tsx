import App from "./App";
import { render, screen } from '@testing-library/react';


describe('App', () => {

  beforeEach(() => { // before each test render the component
    render(<App />);
  });
  
  it('renders correctly', () => {
    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot()
  })

  it('loads main content', () => {
    expect(screen.getByRole("main")).toBeInTheDocument();
  })


})
