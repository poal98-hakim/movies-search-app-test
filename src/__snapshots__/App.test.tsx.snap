// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<body>
  <div>
    <main>
      <div
        class="wrapper-filter"
      >
        <form
          class="form-search flex f-direction-column f-align-center"
        >
          <div
            class="search-bar-wrapper flex f-align-center f-justify-center"
            data-testid="search-bar"
          >
            <img
              alt="icon-search"
              src="search-icon-yellow.png"
              width="25"
            />
            <input
              placeholder="Search*"
              type="text"
              value=""
            />
          </div>
          <div
            class="search-bar-wrapper flex f-align-center f-justify-center"
            data-testid="search-bar"
          >
            <img
              alt="icon-key"
              src="key-icon.png"
              width="25"
            />
            <input
              placeholder="Api Key*"
              type="text"
              value=""
            />
          </div>
          <div
            class="flex f-align-center"
          >
            <input
              class="button"
              type="submit"
              value="Search"
            />
          </div>
        </form>
      </div>
      <div
        class="flex f-justify-center"
      />
    </main>
  </div>
  <div>
    <main>
      <div
        class="wrapper-filter"
      >
        <form
          class="form-search flex f-direction-column f-align-center"
        >
          <div
            class="search-bar-wrapper flex f-align-center f-justify-center"
            data-testid="search-bar"
          >
            <img
              alt="icon-search"
              src="search-icon-yellow.png"
              width="25"
            />
            <input
              placeholder="Search*"
              type="text"
              value=""
            />
          </div>
          <div
            class="search-bar-wrapper flex f-align-center f-justify-center"
            data-testid="search-bar"
          >
            <img
              alt="icon-key"
              src="key-icon.png"
              width="25"
            />
            <input
              placeholder="Api Key*"
              type="text"
              value=""
            />
          </div>
          <div
            class="flex f-align-center"
          >
            <input
              class="button"
              type="submit"
              value="Search"
            />
          </div>
        </form>
      </div>
      <div
        class="flex f-justify-center"
      />
    </main>
  </div>
</body>
`;
